src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+                case DynamicKeywordUseMode.Required:
+                    return currentScopeSeenKeywords.Add(keyword);
+
+                case DynamicKeywordUseMode.RequiredMany:
+                    // Record this because we want to enforce usage when we leave the scope
+                    currentScopeSeenKeywords.Add(keyword);
+                    return true;
+
+                case DynamicKeywordUseMode.OptionalMany:
+                    // We don't care about OptionalMany keywords, so it's more efficient to just move on
+                    return true;
+
+                default:
+                    throw PSTraceSource.NewArgumentException(nameof(keyword.UseMode));
+            }
+        }
+
+        /// <summary>
+        /// Get a list of all the keywords we were required to see in this scope block
+        /// but did not. Note that "required" only checks keywords at this scope level;
+        /// if a keyword from a higher scope is invoked it is not counted. (TODO)
+        /// </summary>
+        /// <returns>a (possibly empty) enumeration of all keywords we "required" the use of</returns>
+        public List<DynamicKeyword> GetUnusedRequiredKeywords()
+        {
+            // If the scope is empty, then we can't require anything
+            if (DynamicKeywordScope.Count == 0)
+            {
+                return null;
+            }
+
+            // If we've run out of "seen" records to check, something has gone wrong
+            if (ScopeSeenDynamicKeywords.Count == 0)
+            {
+                throw new PSInvalidOperationException(String.Format("Tried to pop {0} when it was empty while counting unused required keywords", nameof(ScopeSeenDynamicKeywords)));
+            }
+
+            var unusedKeywords = new List<DynamicKeyword>();
+
+            // Check all the "required" keywords in this scope were seen
+            foreach (var keyword in DynamicKeywordScope.Peek().InnerKeywords.Values)
+            {
+                switch (keyword.UseMode)
+                {
+                    case DynamicKeywordUseMode.Required:
+                    case DynamicKeywordUseMode.RequiredMany:
+                        if (!ScopeSeenDynamicKeywords.Peek().Contains(keyword))
+                        {
+                            unusedKeywords.Add(keyword);
+                        }
+                        break;
+                }
+            }
+
+            return unusedKeywords;
+        }
+
+        /// <summary>
+        /// Wipe this namespace clean. If this is tried while in a scope, we will be in an
+        /// invalid state, hence it is disallowed
+        /// </summary>
+        public void Reset()
+        {
+            if (DynamicKeywordScope.Count > 0)
+            {
+                throw new PSInvalidOperationException("Cannot reset the DynamicKeyword namespace while in a DynamicKeyword scope");
+            }
+
+            _dynamicKeywordScope = new Stack<DynamicKeyword>();
```

Comment:
I think clearing is preferred over creating new instances - the objects will eventually live in Gen2 and there is lower impact on GC.